llpc/translator/lib/SPIRV/SPIRVReader.cpp

```diff
@@ -4074,14 +4080,62 @@ template <> Value *SPIRVToLLVM::transValueWithOpcode<OpExtInst>(SPIRVValue *cons
 
   case SPIRVEIS_Debug:
     return m_dbgTran.transDebugIntrinsic(spvExtInst, block);
-
+  case SPIRVEIS_NonSemanticDebugPrintf:
+    switch (spvExtInst->getExtOp()) {
+    case NonSemanticDebugPrintfDebugPrintf:
+      return transDebugPrintf(spvExtInst, spvArgValues, func, block);
+    default:
+      return nullptr;
+    }
   default:
     llvm_unreachable("Should never be called!");
     return nullptr;
   }
 }
 
 // =====================================================================================================================
+// Handle extended instruction Nonsemantics.debugPrintf
+//
+// @param bi : A SPIR-V debugPrintf instruction
+// @param spvValues: Instruction spirv parameters
+// @param func: Which function to generate code
+// @param bb: Which basicblock to generate code
+Value *SPIRVToLLVM::transDebugPrintf(SPIRVInstruction *bi, const ArrayRef<SPIRVValue *> spvValues, Function *func,
+                                     BasicBlock *bb) {
+  if (!getBuilder()->findResourceNode(ResourceNodeType::DescriptorBuffer, Vkgc::InternalDescriptorSetId,
```

Comment:
I think `DescBuilder` will return poison if the resource node was not found. Maybe you can just skip printf lowering in lgc if the printf buffer pointer is poison. Then we don't need to expose findResourceNode() to Builder anymore.

Reply:
I see the internal discussion, the argument was `InternalDescriptorSetId` usage should only happen in frontend. But now that DescBuilder already has the ability to return poison descriptor if not found, it should be easy to skip printf lowering in lgc pass without using `InternalDescriptorSetId`. The translation happened here is really trivial. Anyway, I don't object to the change if others are ok with this.

Reply:
Originally i want to have a flag from driver to tell llpc to enable/disable debubPrintf. Here i am using the findResourceNode to replacing that flag.  it  should happens as early as possible in the lowering passes and debubPrintf is hence known enabled/disabled explicitly.

Reply:
A flag sounds good to me if we want to skip them early.